File: y22/m07/d19/target_number.py
count = 0


def solution(numbers, target):
    global count
    recursive(numbers, target, 1, 1, 1)
    recursive(numbers, target, 1, -1, 1)
    return count


def recursive(numbers, target, idx, op, value):
    global count
    value += numbers[idx] * op
    if idx == len(numbers) - 1:
        if value == target:
            count += 1
        return
    recursive(numbers, target, idx + 1, 1, value)
    recursive(numbers, target, idx + 1, -1, value)


def solution(numbers, target):
    answer = dfs(numbers, target, 0)
    return answer


def dfs(numbers, target, depth):
    answer = 0
    if depth == len(numbers):
        if sum(numbers) == target:
            return 1
        return 0
    else:
        answer += dfs(numbers, target, depth+1)
        numbers[depth] *= -1
        answer += dfs(numbers, target, depth+1)
        return answer

File: y22/m09/d06/TargetNumber.java
package y22.m09.d06;

public class TargetNumber {
    public static void main(String[] args) {
        Solution solution = new Solution();
//        int[] numbers = new int[]{1, 1, 1, 1, 1};
//        int target = 3;
        int[] numbers = new int[]{4, 1, 2, 1};
        int target = 4;
        int correctAnswer = 2;
        int answer = solution.solution(numbers, target);
        System.out.println(answer == correctAnswer);
        System.out.println("answer = " + answer);
        System.out.println("correctAnswer = " + correctAnswer);
    }
}

class Solution {
    int answer = 0;

    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0, 0);
        return answer;
    }

    public void dfs(int[] numbers, int target, int index, int sum) {
        if (numbers.length == index) {
            if (sum == target) {
                answer++;
            }
            return;
        }
        int number = numbers[index];
        dfs(numbers, target, index + 1, sum + number);
        dfs(numbers, target, index + 1, sum - number);
    }
}

